inc/msp/Client.hpp

```diff
@@ -16,210 +16,364 @@
         //create a shared pointer to a new Subscription and set all properties
         auto subscription = std::make_shared<Subscription<T>>(recv_callback, send_callback, std::make_unique<T>(fw_variant), tp);
-
         //gonna modify the subscription map, so lock the mutex
         std::lock_guard<std::mutex> lock(mutex_subscriptions);
         //move the new subscription into the subscription map
         subscriptions.emplace(id,std::move(subscription));
-
         return subscriptions[id];
     }
 
     /**
-     * @brief hasSubscription check if message ID is subscribed
-     * @param id message ID
-     * @return true if there is already a subscription
-     * @return false if ID is not subscribed
+     * @brief Check if message ID already has a subscription
+     * @param id Message ID
+     * @return True if there is already a matching subscription
      */
     bool hasSubscription(const msp::ID& id) {
         return (subscriptions.count(id)==1);
     }
 
     /**
-     * @brief getSubscription get pointer to subscription
-     * @param id message ID
-     * @return pointer to subscription
+     * @brief Get pointer to subscription
+     * @param id Message ID
+     * @return Shared pointer to subscription (empty if there was no match)
      */
     std::shared_ptr<SubscriptionBase> getSubscription(const msp::ID& id) {
         return subscriptions.at(id);
     }
 
-    void processOneMessage(const asio::error_code& ec,std::size_t bytes_transferred);
-
-
+    /**
+     * @brief Main entry point for processing received data. It
+     * is called directly by the ASIO library, and as such it much match the function 
+     * signatures expected by ASIO.
+     * @param ec ASIO error code 
+     * @param bytes_transferred Number of byte available for processing
+     */
+    void processOneMessage(const asio::error_code& ec, const std::size_t& bytes_transferred);
 
     /**
-     * @brief sendData send raw data and ID to flight controller, accepts any uint8 id
-     * @param id message ID
-     * @param data raw data
+     * @brief Send an ID and payload to the flight controller
+     * @param id Message ID
+     * @param data Raw data (default zero length, meaning no data to send outbound)
      * @return true on success
-     * @return false on failure
      */
-    
     bool sendData(const msp::ID id, const ByteVector &data = ByteVector(0));
-    bool sendData(const msp::ID id, const ByteVector_uptr data)
+    
+    /**
+     * @brief Send an ID and payload to the flight controller
+     * @param id Message ID
+     * @param data Unique pointer to data. May be empty if there is no data to send
+     * @return true on success
+     */
+    bool sendData(const msp::ID id, const ByteVectorUptr&& data)
     {
         //if (!data) std::cout << "ByteVector_uptr is empty" << std::endl;
         if (!data) return sendData(id);
         //std::cout << "sending data from ByteVector_uptr: " << *data << std::endl;
```

Comment:
These comments need to go before the final merge but can stay here for now if you need them for debugging.